### PR TITLE
Interceptors in AoT /4

### LINKED ISSUE #30052: Request Delegate Generator supports interceptors feature
[Request Delegate Generator supports interceptors feature](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-7/#request-delegate-generator-supports-interceptors-feature)

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: 8839fa1e-7c10-4b22-d03d-669906a2b9ea
* Version Independent ID: 8839fa1e-7c10-4b22-d03d-669906a2b9ea
* Content: [ASP.NET Core support for native AOT](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-8.0&viewFallbackFrom=aspnetcore-7.0)
* Content Source: [aspnetcore/fundamentals/native-aot.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/fundamentals/native-aot.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-fundamentals**
* GitHub Login: @mitchdenny
* Microsoft Alias: **midenn**

---
[Associated WorkItem - 147534](https://dev.azure.com/msft-skilling/Content/_workitems/edit/147534)

### LINKED ISSUE #30053: WN: Request Delegate Generator supports interceptors feature
[Request Delegate Generator supports interceptors feature](https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-8-preview-7/#request-delegate-generator-supports-interceptors-feature)

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: f30dbb9e-d64d-a925-01ca-00d9d035e6c1
* Version Independent ID: f30dbb9e-d64d-a925-01ca-00d9d035e6c1
* Content: [What's new in ASP.NET Core 8.0](https://learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-8.0?view=aspnetcore-7.0)
* Content Source: [aspnetcore/release-notes/aspnetcore-8.0.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/release-notes/aspnetcore-8.0.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-releasenotes**
* GitHub Login: @Rick-Anderson
* Microsoft Alias: **riande**

---
[Associated WorkItem - 147533](https://dev.azure.com/msft-skilling/Content/_workitems/edit/147533)

### PR DESCRIPTION
**Pull request by Rick-Anderson**
Fixes #30052
Fixes #30053

Internal review URLs

* [C# 12 Interceptors](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?view=aspnetcore-8.0&branch=pr-en-us-30054#c-12-interceptors)
*  [Increased performance using Interceptors](https://review.learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-8.0?view=aspnetcore-7.0&branch=pr-en-us-30054#increased-performance-using-interceptors)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/fundamentals/native-aot.md](https://github.com/dotnet/AspNetCore.Docs/blob/1a6bbe9e4273f5445be91244069ba4c2c4735593/aspnetcore/fundamentals/native-aot.md) | [ASP.NET Core support for native AOT](https://review.learn.microsoft.com/en-us/aspnet/core/fundamentals/native-aot?branch=pr-en-us-30054) |
| [aspnetcore/release-notes/aspnetcore-8.0.md](https://github.com/dotnet/AspNetCore.Docs/blob/1a6bbe9e4273f5445be91244069ba4c2c4735593/aspnetcore/release-notes/aspnetcore-8.0.md) | [What's new in ASP.NET Core 8.0](https://review.learn.microsoft.com/en-us/aspnet/core/release-notes/aspnetcore-8.0?branch=pr-en-us-30054) |

### CODE REVIEW

File: aspnetcore/release-notes/aspnetcore-8.0.md [hunk 1]
```diff
@@ -235,6 +235,10 @@ The <xref:Microsoft.AspNetCore.Builder.WebApplication.CreateSlimBuilder> method
 
 We've further reduced native AOT binary size for apps that don't need HTTPS or HTTP/3 support. Not using HTTPS or HTTP/3 is common for apps that run behind a TLS termination proxy (for example, hosted on Azure). The new `WebApplication.CreateSlimBuilder` method omits this functionality by default. It can be added by calling `builder.WebHost.UseKestrelHttpsConfiguration()` for HTTPS or `builder.WebHost.UseQuic()` for HTTP/3. For more information, see [The `CreateSlimBuilder` method](xref:fundamentals/native-aot#the-createslimbuilder-method).
 
+### Increased performance using Interceptors
+
+The Request Delegate Generator uses the new [C# 12 interceptors compiler feature](/dotnet/csharp/whats-new/csharp-12) to support intercepting calls to minimal API’s [Map](xref:Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions) action methods with statically generated variants at runtime. The use of interceptors results in increased startup performance for apps compiled with `PublishAot`.
+
```
Comment (Rick-Anderson) [inline aspnetcore/release-notes/aspnetcore-8.0.md:241]: I'd put the table here but it doesn't make sense unless you can disable the interceptors

File: aspnetcore/fundamentals/native-aot.md [hunk 2]
```diff
@@ -246,6 +246,10 @@ A parameter on the delegate that isn't bound to the body and does ***not*** need
 
 :::code language="csharp" source="~/fundamentals/aot/samples/Todo.cs" id="snippet_1":::
 
+### C# 12 Interceptors
+
+The Request Delegate Generator uses the new [C# 12 interceptors compiler feature](/dotnet/csharp/whats-new/csharp-12) to support intercepting calls to minimal API’s [Map](xref:Microsoft.AspNetCore.Builder.EndpointRouteBuilderExtensions) action methods with statically generated variants at runtime. The use of interceptors results in increased startup performance for apps compiled with `PublishAot`.
```
Comment (gfoidl) [inline aspnetcore/fundamentals/native-aot.md:251]: To me "statically generated variants at runtime" reads a bit unlucky, i.e. as these variants are generated at runtime.
Can you please rephrase that? Unfortunately I don't have a proper suggestion for it.

File: aspnetcore/release-notes/aspnetcore-8.0.md (hunk 1 above)
Comment (mitchdenny) [reply aspnetcore/release-notes/aspnetcore-8.0.md:241]: increased or improved?

File: aspnetcore/fundamentals/native-aot.md (hunk 2 above)
Comment (Rick-Anderson) [reply aspnetcore/fundamentals/native-aot.md:251]: @mitchdenny please use the suggestion mechanism

File: aspnetcore/release-notes/aspnetcore-8.0.md (hunk 1 above)
Comment (Rick-Anderson) [reply aspnetcore/release-notes/aspnetcore-8.0.md:241]: @mitchdenny  fixed.

File: aspnetcore/fundamentals/native-aot.md (hunk 2 above)
Comment (mitchdenny) [reply aspnetcore/fundamentals/native-aot.md:251]: Here is how I would write it:

> The Request Delegate Generator uses the new C# 12 interceptors compiler feature to support intercepting calls to minimal APIs Map methods. Map method calls are replaced with code generated calls at compile time. The compiler generated implementations are compatible with native AOT and results in increased startup performance for apps published with the PublishAot property enabled.

**Pull request merged**